### DISCUSSION TITLE
Cannot paginate a dataset

### DISCUSSION THREAD

**Discussion by snacks02** [jeremyevans/sequel#2005] [General]
Hey! I'm a bit lost on this.

This is what I do:

```ruby
Site.dataset.paginate(1, 25)
```

This is what I get:

```
SitesControllerTest#test_#all_sites_table_works:
Sequel::Error: You cannot paginate a dataset that already has a limit
    app/controllers/sites_controller.rb:269:in `all_sites_table'
    test/controllers/sites_controller_test.rb:248:in `block in <class:SitesControllerTest>'
    test/test_helper.rb:23:in `block in run'
    test/test_helper.rb:23:in `run'
```

These are the extensions I have loaded:

```ruby
Sequel::Model.db.extension :pagination
Sequel::Model.db.extension :pg_array
Sequel::Model.db.extension :pg_enum
Sequel::Model.db.extension :pg_json
Sequel::Model.plugin :association_dependencies
Sequel::Model.plugin :validation_helpers
Sequel.extension :migration
```

What might cause this?

**Comment by AlexWayfer** [comment]
> a dataset that already has a limit

Do you have default dataset/filters/limitations in the model? Or maybe spec helpers enables them?

**Reply by snacks02** [reply]
It doesn't happen in `rails console`, but in `test_helper.rb` I only have:

```ruby
# frozen_string_literal: true

require "simplecov"

SimpleCov.start "rails" do
  add_filter "/app/channels/"
end

ENV["RAILS_ENV"] ||= "test"
require_relative "../config/environment"

require "rails/test_help"
require "fixture_dependencies/helper_methods"

class ActiveSupport::TestCase
  parallelize(workers: :number_of_processors)

  include FixtureDependencies::HelperMethods

  FixtureDependencies.fixture_path = "./test/fixtures"

  def run(*args, &)
    Sequel::Model.db.transaction(rollback: :always) { super }
  end
end
```

So I'm not sure what might be causing the issue.

**Reply by AlexWayfer** [reply]
Maybe something in nested files, like `rails/test_help` or `fixture_dependencies/helper_methods`. Maybe parallelization. Not sure either.

You can check the SQL, the exact limit may help you (`dataset.to_sql`, as I remember).

Also you can debug with break-points, even in spec helpers, via `binding.pry` or `binding.irb`, to detect which lines or files affect your default datasets.

**Reply by snacks02** [reply]
Thanks. I tried the suggestions, but they didn't help.

The query is:

```sql
SELECT * FROM "sites"
```

And when I do `.dataset.paginate(1, 25)`, it is:

```sql
SELECT * FROM "sites" LIMIT 25 OFFSET 0
```

And it also doesn't happen when called inside `binding.irb`! Might be something to do with the context where it's called.

**Reply by AlexWayfer** [reply]
> The query is:

I meant to check the query in the problematic place, before `.paginate`. Or maybe higher, upper in stack, you know. For example, in spec helper. If there are no changes to the dataset and no problems with `paginate` in the end of spec helper — yeah, probably something in specific spec file changes it. Some `before` hook? This hook still can be defined somewhere else, but I don't know ways to check when what part of dataset was changed/added, like call stack, so I see opportunity only in the deep debugging, if a place not obvious for you. Also, for example, you can also search for `limit` in all (`spec/`?) files, but this modification still can be even outside of project, what will be visible with the debugging and walking up and down along the stack.

**Comment by snacks02** [comment]
The issue is with [Datagrid](https://github.com/bogdan/datagrid) gem.

I wasn't able to find where it happens, but this results in the error:

```ruby
# all_sites_grid.rb
class AllSitesGrid < BaseGrid
  scope do
    Site
  end
end

# sites_controller.rb
grid = AllSitesGrid.new do
  Site.dataset.paginate(1, 25)
end
```

While this works just fine:

```ruby
# all_sites_grid.rb
class AllSitesGrid < BaseGrid
  scope do
    []
  end
end

# sites_controller.rb
grid = AllSitesGrid.new do
  Site.dataset.paginate(1, 25).to_a
end
```

This is a simplified representation of what was going in my case.
